pkg/trace/otel.go

```diff
@@ -0,0 +1,195 @@
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/getoutreach/gobox/internal/logf"
+	"github.com/getoutreach/gobox/pkg/app"
+	"github.com/getoutreach/gobox/pkg/events"
+	"github.com/getoutreach/gobox/pkg/log"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.10.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type otelTracer struct {
+	Config
+	sync.Once
+	serviceName    string
+	tracerProvider *sdktrace.TracerProvider
+}
+
+// Annotator is a SpanProcessor that adds service-level tags on every span
+type Annotator struct {
+	globalTags GlobalTags
+}
+
+func (a Annotator) OnStart(_ context.Context, s sdktrace.ReadWriteSpan) {
+	setf := func(key string, value interface{}) {
+		s.SetAttributes(attribute.String(key, fmt.Sprintf("%v", value)))
+	}
+
+	logf.Marshal("", app.Info(), setf)
+	logf.Marshal("", a.globalTags, setf)
+}
+
+func (a Annotator) OnEnd(s sdktrace.ReadOnlySpan) {
+	if spanProcessorHook != nil {
+		spanProcessorHook(s.Attributes())
+	}
+}
+
+func (a Annotator) Shutdown(context.Context) error { return nil }
+
+func (a Annotator) ForceFlush(context.Context) error { return nil }
+
+// nolint:gochecknoglobals
+var spanProcessorHook func([]attribute.KeyValue)
+
+// SetPresendHook sets the honeycomb presend hook for testing
+func SetSpanProcessorHook(hook func([]attribute.KeyValue)) {
+	spanProcessorHook = hook
+}
+
+// Deprecated: Use initTracer() instead.
+func (t *otelTracer) startTracing(serviceName string) error {
+	return t.initTracer(context.TODO(), serviceName)
+}
+
+func (t *otelTracer) initTracer(ctx context.Context, serviceName string) error {
+	key, err := t.Otel.APIKey.Data(ctx)
+	if err != nil {
+		log.Error(ctx, "Unable to fetch otel API key", events.NewErrorInfo(err))
+	}
+
+	headers := map[string]string{
+		"x-honeycomb-team":    strings.TrimSpace(string(key)),
+		"x-honeycomb-dataset": t.Otel.Dataset,
+	}
+
+	client := otlptracehttp.NewClient(
+		otlptracehttp.WithEndpoint(t.Otel.Endpoint),
+		otlptracehttp.WithHeaders(headers),
+	)
+
+	exp, err := otlptrace.New(ctx, client)
+	if err != nil {
+		log.Error(ctx, "Unable to start trace exporter", events.NewErrorInfo(err))
+	}
+
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(serviceName),
+		),
+	)
+	if err != nil {
+		log.Error(ctx, "Unable to configure trace provider", events.NewErrorInfo(err))
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exp),
+		sdktrace.WithResource(r),
+		// otel accepts sample rates as a fractions of traces >= 1 will always sample < 0 will never sample
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(t.Otel.SamplePercent/100)),
```

Comment:
Do you know if this uses the same sampling logic as the original Honeycomb sampler?

We need to be careful about changing this logic, since we rely on different services to come to the same conclusions about the same trace IDs.  If they're different, we'll get traces from some services but not others.

Reply:
I was kind of hoping they'd use the same algorithm.  My read is that they're not the same.

- https://github.com/honeycombio/beeline-go/blob/v1.8.0/sample/deterministic_sampler.go
- https://github.com/open-telemetry/opentelemetry-go/blob/sdk/v1.7.0/sdk/trace/sampling.go#L82-L99

They have a similar flavor to them, but the opentelemetry one is using 8 bytes of the trace ID, while the beeline-go one is using 4 bytes.

If I'm right about this, then a service in "opentelemetry mode" will chose to sample a different set of traces than applications in "honeycomb mode".  That's going to make it hard to reconstruct full traces across all our services.